packages/admin/src/form/FinalFormMultiSelect.tsx

```diff
@@ -0,0 +1,114 @@
+
+interface MenuItemProps extends MuiMenuItemProps {
+    children?: React.ReactNode | ((selected: boolean) => React.ReactNode);
+}
+interface FinalFormMultiSelectProps extends FieldRenderProps<string, HTMLDivElement> {
+    children: React.ReactElement<MenuItemProps>[];
+    withSearch?: boolean;
+    inputProps?: InputBaseProps;
+}
+
+function FinalFormMultiSelectComponent({
+    input,
+    classes,
+    children,
+    withSearch = false,
+    inputProps,
+}: WithStyles<typeof styles> & FinalFormMultiSelectProps) {
+    const [searchValue, setSearchValue] = React.useState<string>("");
+    const handleListItemClick = (value: string) => (event: React.SyntheticEvent) => {
+        if (Array.isArray(input.value)) {
+            if (input.value.includes(value)) {
+                if ([...input.value.filter((item) => item !== value)].length === 0) {
+                    input.onChange(undefined);
+                } else {
+                    input.onChange([...input.value.filter((item) => item !== value)]);
+                }
+            } else {
+                if (value) {
+                    input.onChange([value, ...input.value]);
+                } else {
+                    return;
+                }
+            }
+        } else {
+            if (value) {
+                input.onChange([value]);
+            } else {
+                return;
+            }
+        }
+    };
+
+    let items = React.Children.map(children, (child, index) => {
+        if (React.isValidElement(child)) {
+            if (child.props.value && typeof child.props.value === "string") {
+                const selected = input.value?.includes(child.props.value);
+                const children = typeof child.props.children === "function" ? child.props.children(selected) : child.props.children;
+
+                return React.cloneElement(child, {
+                    onClick: handleListItemClick(child.props.value),
+                    selected,
+                    children,
+                });
+            } else {
+                console.error(`child #${index + 1} is missing a value prop`);
+            }
+        } else {
+            console.error(`child #${index + 1} is not a valid ReactElement`);
+        }
+    });
+
+    if (withSearch) {
+        items = items.filter((item) => {
+            if (React.isValidElement(item)) {
+                if (inputProps?.value && inputProps?.value !== "") {
+                    return (item.props.value as string).toLowerCase().includes((inputProps.value as string).toLowerCase());
+                } else if (searchValue !== "") {
+                    return (item.props.value as string).toLowerCase().includes(searchValue.toLowerCase());
+                } else {
+                    return true;
+                }
+            }
+        });
+    }
+
+    return (
+        <>
+            {withSearch && (
+                <InputBase
+                    value={inputProps?.value ? inputProps?.value : searchValue}
```

Comment:
what is inputProps? why should you be able to pass a value here?

Reply:
if you want to control the inputBase from outside :)

Reply:
What would be the use-case or motivation to control the search from outside?

Reply:
if you want to control the input and only allow special characters to type or anything like that :) i just imagined that there will be a case in the future where it can be very usefull :) just tried to make it controllable/flexible as much as possible :)

Reply:
Please remove that possibility, I can't imagine how that can be useful. It adds unneded complexity and additional API.

Reply:
However, we need an API to pass the search value to the outer component (in order to filter the options).

Reply:
I can think of three APIs:

1. FinalFormMultiSelect has a `onSearch?: (value: string) => void` callback. The search value is to stored in the outer component. The options are also filtered in the outer component.
2. FinalFormMultiSelect has a `filterItem?: (item: TItem, searchValue: string) => boolean` prop. The options are filtered in the multi select itself
3. The search field is independent from the FinalFormMultiSelect component. Search value state handling and options filtering is done by the outer component which renders both the input and the select.

Personally I'd vouch for option 3.